cider/exec/operator/join/CiderLinearProbingHashTable.h

```diff
@@ -0,0 +1,284 @@
+    return emplace_impl(value.first.key, std::move(value.second));
+  }
+
+  template <typename... Args>
+  bool insert(Args&&... args) {
+    return emplace_impl(std::forward<Args>(args)...);
+  }
+
+  template <typename... Args>
+  bool emplace(Args&&... args) {
+    return emplace_impl(std::forward<Args>(args)...);
+  }
+
+  // TODO: assert key and value types
+  void merge_other_hashtables(
+      std::vector<std::unique_ptr<LinearProbeHashTable>> otherTables) {
+    int total_size = 0;
+    for (const auto& table_ptr : otherTables) {
+      total_size += table_ptr->size();
+    }
+    rehash(total_size);
+    for (const auto& table_ptr : otherTables) {
+      for (auto it = table_ptr->begin(); it != table_ptr->end(); ++it)
+        insert((*it).first.key, (*it).second);
+    }
+  }
+
+  void swap(LinearProbeHashTable& other) noexcept {
+    std::swap(buckets_, other.buckets_);
+    std::swap(size_, other.size_);
+    std::swap(empty_key_, other.empty_key_);
+  }
+
+  // find
+  folly::fbvector<mapped_type> find(const Key& key) { return find_impl(key); }
+
+  // Bucket interface
+  size_type bucket_count() const noexcept { return buckets_.size(); }
+
+  size_type max_bucket_count() const noexcept { return buckets_.max_size(); }
+
+  // Hash policy
+  void rehash(size_type count) {
+    // comment out due to may adjust load factor in future
+    // count = std::max(count, size() * 2);
+    LinearProbeHashTable other(*this, count);
+    swap(other);
+  }
+
+  void reserve(size_type count) {
+    if (count * 2 > buckets_.size()) {
+      rehash(count * 2);
+    }
+  }
+
+  // Observers
+  hasher hash_function() const { return hasher(); }
+
+  key_equal key_eq() const { return key_equal(); }
+
+ private:
+  template <typename K, typename... Args>
+  bool emplace_impl(const K& key, Args&&... args) {
+    reserve(size_ + 1);
+    for (size_t idx = key_to_idx(key);; idx = probe_next(idx)) {
+      if (!buckets_[idx].first.is_not_null) {
+        buckets_[idx].second = mapped_type(std::forward<Args>(args)...);
+        buckets_[idx].first.key = key;
+        buckets_[idx].first.is_not_null = true;
+        size_++;
+        return true;
+      } else if (key_equal()(buckets_[idx].first.key, key)) {
+        buckets_[idx].first.duplicate_num++;
+      }
+    }
+    return false;
+  }
+
+  template <typename K>
+  folly::fbvector<mapped_type> find_impl(const K& key) {
+    folly::fbvector<mapped_type> vec;
+    int duplicate_num = 0;
+    for (size_t idx = key_to_idx(key);; idx = probe_next(idx)) {
+      if (key_equal()(buckets_[idx].first.key, key) && buckets_[idx].first.is_not_null) {
+        if (vec.size() == 0) {
+          duplicate_num = buckets_[idx].first.duplicate_num;
+        }
+        vec.push_back(buckets_[idx].second);
+        if (duplicate_num == 0) {
+          return vec;
+        }
+      } else if (buckets_[idx].first.is_not_null == false ||
+                 (vec.size() == duplicate_num + unsigned(1))) {
+        return vec;
```

Comment:
Maybe we should handle something instead of just `return vec`?

Reply:
The correctness is test in Ut